src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1231,7 +1231,18 @@ public void RemoveFromWorkspace()
                 // references being converted to metadata references (or vice versa) and we might either
                 // miss stopping a file watcher or might end up double-stopping a file watcher.
                 remainingMetadataReferences = w.CurrentSolution.GetRequiredProject(Id).MetadataReferences;
-                w.OnProjectRemoved(Id);
+                _workspace.RemoveProjectFromTrackingMaps_NoLock(Id);
+
+                // If this is our last project, remove the entire solution so components that expect to see a
+                // SolutionRemoved actually get it
+                if (w.CurrentSolution.ProjectIds.Count == 1)
```

Comment:
not exactly sure how this works.  is this because .CurrentSolution is not updated yet, despite teh call to `_workspace.RemoveProjectFromTrackingMaps_NoLock(Id);` above?  

can we check this info immediately prior to doing any mutation to make things clearer?  or at least doc why this is safe?